### PR TITLE
Some cleanups in Native AOT suspension area

### PR DESCRIPTION
**Pull request by VSadov**
As I was looking through the code to get more familiar, I made some changes. 

These are mostly mechanical changes - renames, removal of dead code, made some frame helpers strongly typed (vs, operating with void*).

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/thread.inl [hunk 1]
```diff
@@ -56,3 +56,18 @@ inline bool Thread::IsInCantAllocStressLogRegion()
 {
     return m_cantAlloc != 0;
 }
+
+struct CantAllocHolder
```
Comment (VSadov) [inline src/coreclr/nativeaot/Runtime/thread.inl:60]: I think, maybe this should be debug-only code. This has now been reduced to ensuring that we do not allocate stresslog buffers in 3 or 4 fairly trivial methods.

We just need to be sure that those methods do not use stress log. I think we could rely on asserts.

File: src/coreclr/nativeaot/Runtime/thread.inl [hunk 2]
```diff
@@ -42,17 +42,40 @@ inline PTR_VOID Thread::GetThreadStressLog() const
...
 {
-    return m_cantAlloc != 0;
+#ifdef _DEBUG
+    return m_forbidBlocking != 0;
+#else
+    return false;
+#endif
 }
+
+struct ForbidBlockingHolder
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.inl:68]: I would stay away from introducing C++ into NativeAOT runtime. In the ideal case, the NativeAOT runtime would not depend on C++ runtime at all. There are not that many places that would have to be cleaned up to make it happen. https://github.com/dotnet/corert/issues/3564 has some old discussion on this.

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/thread.inl:68]: > I would stay away from introducing C++ into NativeAOT runtime. 

I meant introducing constructs that pull in C++ runtime.

Using C++ as "C with classes" is fine.

File: src/coreclr/nativeaot/Runtime/thread.h [hunk 3]
```diff
@@ -27,7 +27,7 @@ class Thread;
 # endif
 #endif // HOST_64BIT
 
-#define TOP_OF_STACK_MARKER ((PTR_VOID)(uintptr_t)(intptr_t)-1)
+static PInvokeTransitionFrame* const TOP_OF_STACK_MARKER = (PInvokeTransitionFrame*)ptrdiff_t(-1);
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.h:30]: Initialized statics in headers tend to produce code bloat. Every compilation unit gets its own copy.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.h:30]: I'd hope they can just elide it as a constant expression, but it can be a define.

File: src/coreclr/nativeaot/Runtime/thread.inl (hunk 2 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.inl:68]: I've seen holders are used in other places, so assumed it is in-style. This particular case can easily do without a holder though.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.inl]: Yes, avoiding c++ runtime and using C-with-classes is reasonable. Some cases are obvious. We do not want `std::this::kind::of::code` or exception handling.

Holders are confusing though. In theory destructors are exception-aware constructs, but I am not sure if they necessarily involve runtime support. It does not look so when I look at emitted assembly. 
We often use destructors just to not write the same code on every exit from a scope.

File: src/coreclr/nativeaot/Runtime/thread.h (hunk 3 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.h]: replaced with a define

File: src/coreclr/nativeaot/Runtime/thread.inl (hunk 2 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/thread.inl]: Methods with on-stack destructors reference C++ exception handling routine in unwind data.

File: src/coreclr/nativeaot/Runtime/stressLog.cpp [hunk 4]
```diff
@@ -132,25 +132,21 @@ void StressLog::Initialize(unsigned facilities,  unsigned level, unsigned maxByt
...
 ThreadStressLog* StressLog::CreateThreadStressLog(Thread * pThread) {
-
     if (theLog.facilitiesToLog == 0)
         return NULL;
 
     if (pThread == NULL)
         pThread = ThreadStore::GetCurrentThread();
 
+    //if we are not allowed to allocate stress log, we should not even try to take the lock
+    _ASSERTE(!pThread->IsInForbidBlockingRegion());
```
Comment (VSadov) [inline src/coreclr/nativeaot/Runtime/stressLog.cpp:142]: This part a bit questionable. When we have a thread in asynchronously interrupted state, we should not do any blocking calls since we do not know what kind of locks the interrupted thread could be holding. 

In theory we need these asserts anyplace where we allocate, do IO or take locks. In practice the "unsafe" state is now very short - it is between the time we suspended the thread and the time we figure if it is in managed code (then we are safe) or not in managed (then we release the thread as not interesting to us). 

Basically we cannot block on the "IsManaged" path, which is very trivial now, so asserting all over the place seems an overkill and asserting only in stresslog seems strange.
Maybe just adding very clear comments on all methods involved would be sufficient?

File: src/coreclr/nativeaot/Runtime/thread.inl (hunk 2 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.inl]: I see. I suspected that. Anyways, I've removed the holder.

File: src/coreclr/nativeaot/Runtime/stressLog.cpp (hunk 4 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/stressLog.cpp:142]: stresslog.cpp/stresslog.h was forked from coreclr for .NET Native UWP. Ideally, we would delete this copy of stresslog.cpp/stresslog.h and switched to using the one from CoreCLR utilcode again (add ifdef to the shared copy around things that are hard to unify).

I can see that some of the issues you have pointed out have been fixed in CoreCLR utilcode copy already.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/stressLog.cpp:142]: I mean more generally - if we track `forbidBlocking` state, we should check for it in a lot of places or not at all. How stresslog is different?

It looks like older code could be trying to do stackwalking without checking `IsManaged` and considered failed stackwalk as `IsManaged` check. That would involve a lot more code while in unsafe state, including stresslog, so the check in stresslog could be just a reaction to that pattern and the bugs that resulted from it.

Considering that the unsafe state is only relevant on the very isolated `IsManaged` code path (basically just an IP range check), I think tracking `forbidBlocking` state maybe not worth it entirely.

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/stressLog.cpp:142]: I agree that this can be probably simplified.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/stressLog.cpp]: I've removed tracking of `forbidBlocking` and just added comments to the few affected methods that should not allocate or take runtime locks.

File: src/coreclr/nativeaot/Runtime/windows/CoffNativeCodeManager.cpp [hunk 5]
```diff
@@ -246,9 +246,8 @@ static_assert(sizeof(CoffNativeMethodInfo) <= sizeof(MethodInfo), "CoffNativeMet
 bool CoffNativeCodeManager::FindMethodInfo(PTR_VOID        ControlPC,
                                            MethodInfo *    pMethodInfoOut)
 {
-    // Stackwalker may call this with ControlPC that does not belong to this code manager
-    if (dac_cast<TADDR>(ControlPC) < dac_cast<TADDR>(m_pvManagedCodeStartRange) ||
-        dac_cast<TADDR>(m_pvManagedCodeStartRange) + m_cbManagedCodeRange <= dac_cast<TADDR>(ControlPC))
+    // Stackwalker may call this with ControlPC that does not managed
+    if (!IsManaged(ControlPC))
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/CoffNativeCodeManager.cpp:250]: Is this a virtual call now?

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/windows/CoffNativeCodeManager.cpp:250]: This is c++, so it might be a direct call, since nothing derives from CoffNativeCodeManager.
It does not have to be virtual though. This can be factored out into something like `NativeCodeManagerBase`. 

There is some other functionality that does not seem to differ between Windows and Unix. I was not sure if we wanted a base class for that.

**Review by jkotas — APPROVED** [review]
Looks good to me otherwise.

File: src/coreclr/nativeaot/Runtime/RuntimeInstance.cpp [hunk 6]
```diff
@@ -66,18 +66,22 @@ PTR_UInt8 RuntimeInstance::FindMethodStartAddress(PTR_VOID ControlPC)
...
 
-ICodeManager * RuntimeInstance::FindCodeManagerByAddress(PTR_VOID pvAddress)
+// WARNING: This method is called by suspension while one thread is interrupted
+//          in a random location, possibly holding random locks.
+//          It is unsafe to use blocking APIs or allocate in this method.
+//          Please ensure that all methods called by this one also have this warning.
+bool RuntimeInstance::IsManaged(PTR_VOID pvAddress)
 {
-    ReaderWriterLock::ReadHolder read(&m_ModuleListLock);
+    return m_CodeManager->IsManaged(pvAddress);
```
[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/RuntimeInstance.cpp:75]: I think ability to check if an IP is in managed code is important. The check should be fast and simple (since we can't take locks or allocate). The original implementation was walking a linked list while taking a read lock.

If we would want to support multiple code managers, I'd prefer a model when adding another code manager merges its data into some lookup data structure that is optimized for lookup.

I think that would be like moving `IsManaged` API to the `RuntimeInstance` and for now implementing it by storing the range in the instance once a code manager is added..

[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/RuntimeInstance.cpp:75]: > If we would want to support multiple code managers, I'd prefer a model when adding another code manager merges its data into some lookup data structure that is optimized for lookup.

Yes, that's fine. The knowledge of the address ranges managed by each code manager is still in the runtime instance in this model.
